### PR TITLE
Adding the 'ClassDesigner' capability to the C# and VB design time targets.

### PR DESCRIPTION
**Pull request by tannergooding**
FYI. @Pilchie, @davkean.

This resolves https://github.com/dotnet/project-system/issues/256.

I've also tagged you on the CPS side change.

### CODE REVIEW

**Comment by davkean** [comment]
Can we separate the adding the capability from the formatting changes?

**Comment by davkean** [comment]
This change is being made to 15.5 but the CPS change is being made to 15.6, please make them to the same branch.

**Comment by tannergooding** [comment]
@davkean, as far as I can tell, CPS never renamed their branches.

**Comment by davkean** [comment]
In CPS, master is 15.5. v15.6 is 15.6.

**Comment by tannergooding** [comment]
@davkean, I will update then. This also means I was told the wrong place to target for the DisplayOrder changes and we need to backport them to master from v15.6

**Comment by tannergooding** [comment]
@davkean, you can append `?w=1` to the end of the URL to hide whitespace changes (https://github.com/dotnet/project-system/pull/2817/files?w=1). CodeFlow has a button that provides similar functionality.

It is significantly more effort to split apart whitespaces changes (especially changes that are automatically made by the formatting engine in the editor) that it is to review the PR with them ignored. It also means that they are less likely to get fixed in the future, as well as less likely to be reviewed over actual code changes.

**Comment by tannergooding** [comment]
The other bug called out appears to only be when opening a class diagram that is included as a solution item and is marked as being "already open" by the suo file (so that it will be automatically opened by the editor when the solution is opened again).

Adding class diagrams to a solution and opening them in general is not working with this change due to the project capability not being there.

**Comment by Andrey-iTech** [comment]
@tannergooding, you a right. If remove .suo file (or just close diagram before closing the solution), the message box doesn't appear.

**Comment by tannergooding** [comment]
Split the formatting change into a separate commit.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets [hunk 1]
```diff
@@ -25,7 +25,7 @@
     <ItemGroup>
      <PropertyPageSchema Include="$(ManagedXamlResourcesDirectory)CSharp.ProjectItemsSchema.xaml;"/>
 
-     <ProjectCapability Include="CSharp;Managed"/>
+     <ProjectCapability Include="CSharp;Managed;ClassDesigner;"/>
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets]: Don't put a trailing ";"

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.VisualBasic.DesignTime.targets [hunk 2]
```diff
@@ -30,7 +30,7 @@
         <Context>ProjectSubscriptionService</Context>
       </PropertyPageSchema>
 
-     <ProjectCapability Include="VB;Managed"/>
+     <ProjectCapability Include="VB;Managed;ClassDesigner;"/>
   </ItemGroup>
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.VisualBasic.DesignTime.targets]: Don't put a trailing ";"

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets [hunk 3]
```diff
@@ -22,25 +22,27 @@
...
 
-    <ItemGroup>
-     <PropertyPageSchema Include="$(ManagedXamlResourcesDirectory)CSharp.ProjectItemsSchema.xaml;"/>
+  <ItemGroup>
+    <PropertyPageSchema Include="$(ManagedXamlResourcesDirectory)CSharp.ProjectItemsSchema.xaml;"/>
 
-     <ProjectCapability Include="CSharp;Managed"/>
-    </ItemGroup>
+    <ProjectCapability Include="
+                          CSharp;
```
Comment (davkean) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets]: Can you make sure that CPS removes newline when it calculates capabilities?

**Comment by davkean** [comment]
@tannergooding Okay, let's not block this check-in on that other bug - but we should file a bug for it.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets (hunk 1 above)
Comment (tannergooding) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets]: We put a trailing ; in our other targets. I was making this consistent.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets (hunk 3 above)
Comment (tannergooding) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.CSharp.DesignTime.targets]: We already do this in our other DesignTime targets.

**Comment by tannergooding** [comment]
FYI. @Pilchie for ask mode approval.

**Comment by Pilchie** [comment]
@davkean - can you take another look?

**Comment by davkean** [comment]
Can we get a bug filed on that open document bug?

**Comment by tannergooding** [comment]
I'll get one filed before I merge tonight.

**Comment by davkean** [comment]
Please don't merge this until we have the CPS changes in and confirmed.

**Comment by tannergooding** [comment]
@davkean, trying this out on the latest bits, I don't see any warning dialogs when attempting to open or add class diagram files after the solution has been loaded.

However, if a class diagram is set as one of the "opened" files in the suo (close solution with the file in an active tab, and reopen the solution so the file is automatically opened) we get a NRE.

This is because there is a call to `string CodeHandler.GetRootNamespaceName(out string language)` which attempts to get the `CodeModel` from the `EnvDTE.Project` and our `GetCodeModel` implementation (https://github.com/dotnet/project-system/blob/master/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/ProjectContextCodeModelProvider.cs#L39) returns null if the language service has not created an `ActiveProjectContext` yet.

**Comment by davkean** [comment]
I filed https://github.com/dotnet/project-system/issues/2998 to represent the language service init issue.

**Pull request merged**